File: Backend/src/main/java/com/CrowdHaven/Backend/services/PostService.java
package com.CrowdHaven.Backend.services;

import com.CrowdHaven.Backend.DTOS.PostDTO;
import com.CrowdHaven.Backend.models.*;
import com.CrowdHaven.Backend.repositories.CommunityRepository;
import com.CrowdHaven.Backend.repositories.MemberCommunityRepository;
import com.CrowdHaven.Backend.repositories.PostRepository;
import com.CrowdHaven.Backend.repositories.UserRepository;
import lombok.AllArgsConstructor;
import org.springframework.stereotype.Service;

import java.util.List;
import java.util.Optional;

@Service
@AllArgsConstructor
public class PostService {

    private final UserRepository userRepository;
    private final CommunityRepository communityRepository;
    private final PostRepository postRepository;
    private final MemberCommunityRepository memberCommunityRepository;

    public List<Post> getAllPosts() {
        return postRepository.findAll();
    }

   public List<Post> getPostsByCommunity(Long communityId) {
        return postRepository.findByCommunityId(communityId);
   }

    public List<Post> getPostsByUser(Long userId) {
        return postRepository.findByUserId(userId); // Llama al repositorio para obtener los productos
    }

    public Optional<Post> getPostById(Long postId) {
        return postRepository.findById(postId);
    }


    public Post createPost(PostDTO postFromFront) {
        // Validar usuario
        User user = userRepository.findById(postFromFront.getUserId())
                .orElseThrow(() -> new IllegalArgumentException("Usuario no encontrado"));

        // Validar comunidad
        Community community = communityRepository.findById(postFromFront.getCommunityId())
                .orElseThrow(() -> new IllegalArgumentException("Comunidad no encontrada"));

        // Validar que el usuario pertenezca a la comunidad
        boolean isMember = memberCommunityRepository
                .findByUserIdAndCommunityId(user.getId(), community.getId())
                .isPresent();

        if (!isMember) {
            throw new IllegalArgumentException("El usuario no pertenece a la comunidad");
        }

        Post post = new Post();
        post.setUser(user);
        post.setCommunity(community);
        post.setTitle(postFromFront.getTitle());
        post.setContent(postFromFront.getContent());
        post.setImage(postFromFront.getImage());

        post = postRepository.save(post);

        return post;
    }


    public void deletePost(Long id) {
        postRepository.deleteById(id);
    }

    public Post updatePostReaction(Long id, boolean isLike) {

        Optional<Post> existingPost = postRepository.findById(id);

        if (existingPost.isEmpty()) {
            throw new IllegalArgumentException("Comentario no encontrado");
        }

        Post post = existingPost.get();

        if (isLike) {
            post.setLike_count(post.getLike_count() + 1);
        } else {
            post.setDislike_count(post.getDislike_count() + 1);
        }

        return postRepository.save(post);
    }
}

File: CrowdHaven-front/src/app/api/services/post/post.service.ts
import { Injectable } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { Observable } from 'rxjs';
import { Post } from '../../models/post.model';
import { PostDTO } from '../../dtos/post-dto';

@Injectable({
  providedIn: 'root',
})
export class PostService {
  private apiUrl = 'http://localhost:8080/CrowdHaven/posts';

  constructor(private http: HttpClient) {}

  getAllPosts(): Observable<Post[]> {
    return this.http.get<Post[]>(this.apiUrl);
  }

  getPostById(id: number): Observable<Post> {
    return this.http.get<Post>(`${this.apiUrl}/${id}`);
  }

  getPostsByCommunity(communityId: number): Observable<Post[]> {
    return this.http.get<Post[]>(`${this.apiUrl}/community/${communityId}`);
  }

  getPostsByUser(userId: number): Observable<Post[]> {
    return this.http.get<Post[]>(`${this.apiUrl}/user/${userId}`);
  }

  createPost(postDTO: PostDTO): Observable<Post> {
    return this.http.post<Post>(this.apiUrl, postDTO);
  }

  updatePost(postId: number, PostDTO: PostDTO): Observable<Post> {
    return this.http.put<Post>(`${this.apiUrl}/${postId}`, PostDTO);
  }

  deletePost(id: number): Observable<void> {
    return this.http.delete<void>(`${this.apiUrl}/${id}`);
  }

  updatePostReaction(id: number, isLike: boolean): Observable<Post> {
    return this.http.put<Post>(`${this.apiUrl}/${id}/reaction?isLike=${isLike}`, {});
  }

  likePost(postId: number): Observable<void> {
    return this.http.post<void>(`${this.apiUrl}/${postId}/like`, {});
  }

  dislikePost(postId: number): Observable<void> {
    return this.http.post<void>(`${this.apiUrl}/${postId}/dislike`, {});
  }
}
